docs/coding-guidelines.md

```diff
@@ -0,0 +1,269 @@
+    - [Keep things sequential unless proven to be slow](#keep-things-sequential-unless-proven-to-be-slow)
+    - [Use `async/await` for sequential execution only](#use-asyncawait-for-sequential-execution-only)
+    - [Don't communicate by sharing memory; share memory by communicating.](#dont-communicate-by-sharing-memory-share-memory-by-communicating)
+        - [Further Reading](#further-reading)
+
+<!-- markdown-toc end -->
+
+
+Below is a set of coding guidelines followed across the rust-libp2p code base.
+
+## Hierarchical State Machines
+
+If you sqint, rust-libp2p is just a big hierarchy of [state
+machines](https://en.wikipedia.org/wiki/Finite-state_machine) where parents pass
+events down to their children and children pass events up to their parents.
+
+<details>
+    <summary>Code</summary>
+
+    ```
+    @startuml
+    Swarm <|-- RootBehaviour
+    Swarm <|-- ConnectionPool
+    Swarm <|-- Transport
+    RootBehaviour <|-- PingBehaviour
+    RootBehaviour <|-- IdentifyBehaviour
+    RootBehaviour <|-- KademliaBehaviour
+
+    Swarm : poll()
+    RootBehaviour : poll()
+    ConnectionPool : poll()
+    Transport : poll()
+    PingBehaviour : poll()
+    IdentifyBehaviour : poll()
+    KademliaBehaviour : poll()
+    @enduml
+    ```
+</details>
+
+```
+                                         ,------.
+                                         |Swarm |
+                                         |------|
+                                         |poll()|
+                                         `------'
+                                             |
+                                             |
+                   ,-------------.   ,--------------.   ,---------.
+                   |RootBehaviour|   |ConnectionPool|   |Transport|
+                   |-------------|   |--------------|   |---------|
+                   |poll()       |   |poll()        |   |poll()   |
+                   `-------------'   `--------------'   `---------'
+                           |
+,-------------.  ,-----------------.   ,-----------------.
+|PingBehaviour|  |IdentifyBehaviour|   |KademliaBehaviour|
+|-------------|  |-----------------|   |-----------------|
+|poll()       |  |poll()           |   |poll()           |
+`-------------'  `-----------------'   `-----------------'
+```
+
+Using hierarchical state machines is a deliberate choice throughout the
+rust-libp2p code base. It makes reasoning about control and data flow easy and
+works well with Rust's `Future` model. The archetecture pattern of
+hierarchical state machines should be used wherever possible.
+
+### Conventions for `poll` implementations
+
+The `poll` method of a single state machine can be complex especially when that
+state machine itself `poll`s many child state machines. The patterns shown below
+have proven useful and should be followed across the code base.
+
+``` rust
+fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output>{
+  loop {
```

Comment:
Make a comment here about looping vs recursing.

Reply:
:+1: Added in 3b07e71c.